File: apps/frontend/src/adapters/shared/scenarios.ts
import type {
  NewPortfolioScenario,
  PortfolioScenario,
  ScenarioPerformanceResult,
} from "@/lib/types";
import { invoke, logger } from "./platform";

export const getScenarios = async (): Promise<PortfolioScenario[]> => {
  try {
    return await invoke<PortfolioScenario[]>("get_scenarios");
  } catch (error) {
    logger.error("Error fetching scenarios.");
    throw error;
  }
};

export const getScenario = async (scenarioId: string): Promise<PortfolioScenario> => {
  try {
    return await invoke<PortfolioScenario>("get_scenario", { scenarioId });
  } catch (error) {
    logger.error("Error fetching scenario.");
    throw error;
  }
};

export const getScenarioPerformance = async (
  scenarioId: string,
  startDate?: string,
  endDate?: string,
): Promise<ScenarioPerformanceResult> => {
  try {
    return await invoke<ScenarioPerformanceResult>("calculate_scenario_performance", {
      scenarioId,
      startDate,
      endDate,
    });
  } catch (error) {
    logger.error("Error replaying scenario performance.");
    throw error;
  }
};

export const createScenario = async (
  scenario: NewPortfolioScenario,
): Promise<PortfolioScenario> => {
  try {
    return await invoke<PortfolioScenario>("create_scenario", { scenario });
  } catch (error) {
    logger.error("Error creating scenario.");
    throw error;
  }
};

export const updateScenarioEntry = async (
  scenario: PortfolioScenario,
): Promise<PortfolioScenario> => {
  try {
    const payload: NewPortfolioScenario = {
      name: scenario.name,
      description: scenario.description,
      kind: scenario.kind,
      accountScope: scenario.accountScope,
      asOfDate: scenario.asOfDate,
      benchmarkSymbols: scenario.benchmarkSymbols,
      basket: scenario.basket,
      assumptions: scenario.assumptions,
    };
    return await invoke<PortfolioScenario>("update_scenario_entry", {
      scenarioId: scenario.id,
      scenario: payload,
    });
  } catch (error) {
    logger.error("Error updating scenario.");
    throw error;
  }
};

export const deleteScenario = async (scenarioId: string): Promise<void> => {
  try {
    await invoke<void>("delete_scenario_entry", { scenarioId });
  } catch (error) {
    logger.error("Error deleting scenario.");
    throw error;
  }
};

File: apps/tauri/src/commands/scenarios.rs
use std::sync::Arc;

use tauri::State;

use crate::context::ServiceContext;
use wealthfolio_agent_tools::{replay_scenario_performance, SymbolPerformanceOutput};
use wealthfolio_core::scenarios::{NewPortfolioScenario, PortfolioScenario};

#[tauri::command]
pub async fn get_scenarios(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<PortfolioScenario>, String> {
    state
        .scenario_service()
        .list_scenarios()
        .map_err(|e| format!("Failed to load scenarios: {}", e))
}

#[tauri::command]
pub async fn get_scenario(
    scenario_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PortfolioScenario, String> {
    state
        .scenario_service()
        .get_scenario(&scenario_id)
        .map_err(|e| format!("Failed to load scenario: {}", e))
}

#[tauri::command]
pub async fn create_scenario(
    scenario: NewPortfolioScenario,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PortfolioScenario, String> {
    state
        .scenario_service()
        .create_scenario(scenario)
        .await
        .map_err(|e| format!("Failed to create scenario: {}", e))
}

#[tauri::command]
pub async fn update_scenario_entry(
    scenario_id: String,
    scenario: NewPortfolioScenario,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<PortfolioScenario, String> {
    state
        .scenario_service()
        .update_scenario(&scenario_id, scenario)
        .await
        .map_err(|e| format!("Failed to update scenario: {}", e))
}

/// Replay a saved basket scenario over history and return its return series,
/// shaped so the Performance page can plot it as a benchmark reference line.
#[tauri::command]
pub async fn calculate_scenario_performance(
    scenario_id: String,
    start_date: Option<String>,
    end_date: Option<String>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<SymbolPerformanceOutput, String> {
    replay_scenario_performance(
        state.agent_environment(),
        &scenario_id,
        start_date.as_deref(),
        end_date.as_deref(),
    )
    .await
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_scenario_entry(
    scenario_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<(), String> {
    state
        .scenario_service()
        .delete_scenario(&scenario_id)
        .await
        .map_err(|e| format!("Failed to delete scenario: {}", e))
}
